File: src/main/java/chapter5/schedulers/ComputationSchedulerExample.java
package chapter5.schedulers;

import common.CommonUtils;
import common.Log;
import io.reactivex.Observable;

import java.util.concurrent.TimeUnit;


public class ComputationSchedulerExample {

    public void basic() {
        CommonUtils.exampleStart();

        String[] objs = {"1", "3", "5"};
        Observable<String> source = Observable.fromArray(objs)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS), (a, b) -> a);

        source.map(item -> "<<" + item + ">>")
                .subscribe(Log::it);

        source.map(item -> "##" + item + "##")
                .subscribe(Log::it);

        CommonUtils.sleep(500);
    }

    public static void main(String[] args) {
        ComputationSchedulerExample demo = new ComputationSchedulerExample();
        demo.basic();
    }

}

File: src/main/kotlin/chapter5/schedulers/ComputationSchedulerExample_.kt
package chapter5.schedulers

import common.CommonUtils_
import common.Log_
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import java.util.concurrent.TimeUnit


class ComputationSchedulerExample_ {

    fun basic() {
        CommonUtils_.exampleStart()

        val objs = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*objs)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS),
                        BiFunction { t1: String, _: Long -> t1 })

        source.map { "<<$it>>" }
                .subscribe { Log_.it(it) }

        source.map { "##$it##" }
                .subscribe { Log_.it(it) }

        CommonUtils_.sleep(500)
    }

}


fun main(args: Array<String>) {
    val demo = ComputationSchedulerExample_()
    demo.basic()
}
